### ISSUE TITLE
Python Debugger: "Timed out waiting for launcher to connect" 

### ISSUE THREAD

**Issue by MJ1Peter** [microsoft/debugpy#1677]
Having an issue working in VSCode (Version: 1.90.0) using Python 3.12.2 on Mac. Program (below) runs just fine from the terminal:

Program
```
import sys

print (len(sys.argv))
print (sys.argv)
if  len(sys.argv) < 3:
    print("specify <sheet-name> <operation> where operation is 'create', 'insert', show' or 'delete'")
    sys.exit(1)
        
print("argument list:", str(sys.argv))
sheetname =  sys.argv[1]
operation =  sys.argv[2]
```

From terminal
```
petermelo@Peters-MacBook-Pro Code % python3 -V 
Python 3.12.2
petermelo@Peters-MacBook-Pro Code % python3 Test3.py 
1
['Test3.py']
specify <sheet-name> <operation> where operation is 'create', 'insert', show' or 'delete'
petermelo@Peters-MacBook-Pro Code % 
```

However, when I try to debug it from VSCode I get the following: 
![Image 6-8-24 at 10 19 AM](https://github.com/microsoft/debugpy/assets/162541639/2d84e740-c47c-4a71-9ef0-9b488351278c)

This is the config for launch.json im using: 

`{
    
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    
    "version": "0.2.0",
    
    "configurations": [
              {
                "name": "Python Debugger: Current File (Integrated Terminal)",
                "type": "debugpy",
                "request": "launch",
                "program": "${file}",
                "console": "integratedTerminal"
              }
    ]
}`

**Comment by iliagerman** [comment]
Any updates?

**Comment by TheJedinator** [comment]
Same issue for me changing to internalConsole works.

**Comment by josephkirankk** [comment]
Any updates ?

**Comment by nleroy917** [comment]
related? https://github.com/microsoft/debugpy/issues/1681

**Comment by neeox** [comment]
Same issue for me, works with `internalConsole` as well.

**Comment by rchiodo** [comment]
This seems like a vscode extension issue, not a debugpy issue. Debugpy doesn't pass the command to the terminal.

**Comment by rchiodo** [comment]
@karthiknadig I think this would be a problem with the extension (or maybe VS code) running the command before the terminal starts up.

**Comment by karthiknadig** [comment]
@rchiodo This is still with `debugpy`. This is the sequence of communication that occurs for a typical launch configuration. 

There are 4 process involved. 
P1 => VS Code 
P2 => DAP (debugpy.adapter)
P3 => Debugpy launcher
P4 => User code with debug.server injected into it.

Sequence:
1. User initiates a launce (by debug UI or Test UI).
2. Python extension (in VS Code P1) generates the DAP configuration and starts the DAP server process P2. This  `type: launch` configuration uses stdio for DAP communication.
3. Then P1 and P2 go through the Initialization sequence. (see https://github.com/microsoft/debugpy/wiki/DAP-Client-reference#dap-implementation-details)
4. After receiving the launch configuration. P2 opens a port and generates a command to return with `run-in-terminal` request with the port number. This is for the `integratedTerminal` console case. For other case P3 is launched directly by P2.
5. VS Code P1 receives this command via `run-in-terminal` request, and creates a new terminal for debugging, and starts P3 from the terminal.
6. P3 then connects back to P2.  (<=== this is where the issue is occurring) as far as I can tell.
7. After connecting P3 receives the User code details, like environment variables, and other config, like wait on exit, and launches user code with debug.server module injected into the process (P4).
8. P4 Also connects back to P2. From this point, any subprocess started by P4 is also wired to communicate with P2. P2 acts as the main coordinator that receives commands from VS Code via DAP and it sends it across to P4 and any subprocesses.

I believe the issue occurs at 6. I am guessing something is going on with how the port is setting up.

The reason we need P3 and we don't directly launch P4 is to handle OS exit in Python. Typically, when python process exits we use `sys.exit`, and we can handle that with inproc debug server. But OS exist, does not allow us any opportunity to handle exit inproc.

Here is the code in P1 that starts P2:
https://github.com/microsoft/vscode-python-debugger/blob/afd0d2bbb634e51064806daa013fef5895d6c52e/src/extension/debugger/adapter/factory.ts#L38-L100

**Comment by wanggaa** [comment]
Im also having the same timeout issue, and the temporary fix `change in a launch.json: "console": "internalConsole"` do not work for me.

**Comment by mcrysler-ng** [comment]
Any updates?

**Comment by rchiodo** [comment]
We have yet to reproduce the problem. I still think it's some issue with VS code as debugpy doesn't change very often. You might try an older VS code and see if the problem goes away.  That would be step 5 above. The 'run-in-terminal' message isn't starting the P3 process.

**Comment by Sube-py** [comment]
It's finally fixed😄
```txt
1.94.0
d78a74bcdfad14d5d3b1b782f87255d802b57511
arm64
```

**Comment by jjwwczy** [comment]
**Solved.** I am using version 1.94.0 on Windows.

Right-click on Code.exe and select "**Properties**" from the context menu.
In the Properties window, switch to the "**Compatibility**" tab.
Uncheck the box that says "**Run this program in compatibility mode fo****r.**"
I found that my PC had this option **checked**. The problem was resolved after I **unchecked** it.

**Comment by wabiloo** [comment]
I still have the same issue on Mac, even with latest version...

**Comment by Sube-py** [comment]
> I still have the same issue on Mac, even with latest version...

Did you use conda?

**Comment by wabiloo** [comment]
For what?  But no, I've never used conda

**Comment by mwdiers** [comment]
This is still a problem in 1.99.3. If using integratedTerminal, the first time I launch a session, it times out. Thereafter it no longer times out. It does seem to be sending the command before the terminal is ready.

As before, switching to internalConsole solves the problem, but that is not ideal.

**Comment by FIFER-Mods** [comment]
Yes, I still run into this problem on 1.96.2

**Comment by andyatshippo** [comment]
Running into this problem as well
pydebug: 2025.8.0

VSCode:
Version: 1.100.3 (Universal)
Commit: 258e40fedc6cb8edf399a463ce3a9d32e7e1f6f3
Date: 2025-06-02T13:30:54.273Z (1 wk ago)
Electron: 34.5.1
ElectronBuildId: 11369351
Chromium: 132.0.6834.210
Node.js: 20.19.0
V8: 13.2.152.41-electron.0
OS: Darwin arm64 24.4.0

**Comment by rchiodo** [comment]
Given that this issue is closed, you may be experiencing different circumstances, I'd recommend opening a new issue. I believe the root cause of this specific problem was VS code not actually running the terminal commands. If you see the debugpy commands in your terminal, then you likely have a different issue.

**Comment by iliagerman** [comment]
I don't think it was actually resolved, it keeps happening

**Comment by rchiodo** [comment]
- > I don't think it was actually resolved, it keeps happening

We'd need logs and a lot more detail to debug what the problem might be. Opening a new issue automatically collects some of that information but if people want to provide it here, that works too.

**Comment by rchiodo** [comment]
This all works fine for me. Latest VS Code stable, latest debugger:

![Image](https://github.com/user-attachments/assets/12b2ddd8-19b9-49ca-a522-ecd38b66b906)

**Comment by tejasavora** [comment]
I am on MacOS, and was happening while running tests. I added following to `settings.json` and it works fine now.
`"python.terminal.activateEnvironment": false`

**Comment by BohdanKostivKrootl** [comment]
Hey, long time no see — 10 months after my previous comment in this thread :)
The problem reoccurred, after a VS Code update, of course.

And yeah, the solution with changing "console" configuration parameter to "internalConsole" or "externalTerminal" still works, so it's the same issue I guess. Hope it will be fixed soon

**Comment by wabiloo** [comment]
Same problem for me. And just like last time, the workaround does not work for me

— Fabre

**Comment by karanbalani** [comment]
Same issue is happening for me, probably after the recent update.
And the temporary fix using `"console": "internalConsole"` in `launch.json` works for me.

**Comment by tsimader** [comment]
> I am on MacOS, and was happening while running tests. I added following to `settings.json` and it works fine now. `"python.terminal.activateEnvironment": false`

Worked for me!

**Comment by karanbalani** [comment]
@sito115 @tejasavora - I think in most of the cases, people would be working with a virtual environment, and turning `python.terminal.activateEnvironment` off, would need the deps to be installed globally, which IMO would not be an idea setup.

**Comment by tsimader** [comment]
@balanikaran - Good point. In my case, VS Code still recognizes the virtual environment because I is defined in the "Select Interpreter" option. I am not sure if it is the best solution, but at least it does not show the "Timed out wating for launcher to connect" error.

**Comment by sqzw-x** [comment]
I think this problem may not about debugpy, but there are indeed two or more extensions racing for the python debug console.
I uninstalled the `python-environments` extension and solved the problem. Since this issue is earlier than that extension, I don't think this is the root cause.
In my case, The python-environments extension sent `source xxx/.venv/bin/activate` cmd to activate the venv **AFTER** the debug cmd (`/usr/bin/env xxx/.venv/bin/python xxx/.vscode/extensions/ms-python.debugpy-2025.10.0-darwin-arm64/bundled/libs/debugpy/adapter/../../debugpy/launcher 59912 -- xxx/main.py`), and broke the debug cmd. After I uninstalled it, the `python` extension would send the same command, but **BEFORE** the debug cmd, making the problem resoved.

**Comment by diggy128** [comment]
Fixed for a while by disabling Python Environments and rolling back to python 2025.18.0
Still occurring after updating to python 2026.4.0 and python debugger 2026.4.0
AND after updating to python debugger 2026.6.0

Workaround as mentioned: "console": "internalConsole",

openSUSE Leap 16, KDE plasma
VS Code API Version: 1.110.1
